### ISSUE TITLE
Custom expression function Median() fails against SQL Server

### ISSUE THREAD

**Issue by notrom** [metabase/metabase#12222]
**Edit:** Just finished reading the docs for expressions and noted the limitations section listing these functions. Going to leave it open though, please consider it a feature request :)

**Describe the bug**
Custom expression Median() is not working against SQL Server data sources. Get an error in the notebook editor when previewed.

I'm so looking forward to this working, we use quite a few median's in our reporting.

**Logs**
Nothing in either the javascript console or the server logs.

**To Reproduce**
1. Start a new question against a data source from SQL Server
2. Summarise using a Custom Expression
3. Expression "Median([number field])", give it a name
4. Hit the preview button
5. Error displayed in the preview pane: "The function 'PERCENTILE_CONT' must have an OVER clause."

**Expected behavior**
The median is calculated and presented to the user

**Screenshots**
![image](https://user-images.githubusercontent.com/4504437/78098202-d1806800-743a-11ea-96d2-9fe8f667d5bb.png)

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "Cp1252",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "13.0.2+8",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "https://java.oracle.com/",
    "java.version": "13.0.2",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "13.0.2+8",
    "os.name": "Windows 10",
    "os.version": "10.0",
    "user.language": "en",
    "user.timezone": "Pacific/Auckland"
  },
  "metabase-info": {
    "databases": [
      "sqlserver",
      "postgres",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "12.1"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-03-31",
      "tag": "v0.35.0",
      "branch": "master",
      "hash": "3f96c47"
    },
    "settings": {
      "report-timezone": "Pacific/Auckland"
    }
  }
}
```

**Severity**
This is a new function so doesn't break anything for us but it was one we were very excited to see coming, shame it's not working for us. We can continue to calculate medians outside of Metabase but it's a frustration.

**Additional context**
The Median() function seems to work fine against the Sample Dataset, it seems to be a SQL Server syntax issue.

**Comment by notrom** [comment]
Same with Percentile() as it uses the same underlying SQL command.

**Comment by flamber** [comment]
@notrom The documentation limitations is specific to MySQL, so either it needs to get updated or this might actually be a bug.
https://www.metabase.com/docs/latest/users-guide/expressions.html#database-limitations
Which version of SQL Server are you using?

**Comment by sbelak** [comment]
Documentation didn't get updated. As far as I can tell MSSQL does not support median outside analytic/window queries which is too limiting for how we use aggregations.

**Comment by camsaul** [comment]
So, we'd want to support `:percentile` as well here since they're both feature flagged under the same feature in Metabase drivers. If we can use a Percentile function somewhere, then we can use that to calculate median if there is not an explicit median function.

The SQL Server link I posted above is actually for MDX rather than normal SQL and I don't really have a good sense for whether or not that means we'd be able to leverage it. Either way, it looks like they support what looks at a glance like a window function [`PERCENTILE_DISC`](https://learn.microsoft.com/en-us/sql/t-sql/functions/percentile-disc-transact-sql?view=sql-server-ver16), also `PERCENTILE_CONT`, and some aggregate versions like [`APPROX_PERCENTILE_DISC`](https://learn.microsoft.com/en-us/sql/t-sql/functions/approx-percentile-disc-transact-sql?view=sql-server-ver16). But the syntax looks a little wierd:

```sql
APPROX_PERCENTILE_DISC (numeric_literal)  
WITHIN GROUP (ORDER BY order_by_expression [ASC|DESC]
```

example:

```sql
SELECT DeptId,
APPROX_PERCENTILE_DISC(0.10) WITHIN GROUP(ORDER BY Salary) AS 'P10',
APPROX_PERCENTILE_DISC(0.90) WITHIN GROUP(ORDER BY Salary) AS 'P90'
FROM tblEmployee
GROUP BY DeptId
```

Seems _possible_ to make this work but I'm not sure if having to copy over the `ORDER BY` we're doing elsewhere would make this too hairy or not

**Comment by lbrdnk** [comment]
Hi @camsaul, I'm interested in implementing this. First I wanted to do it in [#yolo](https://www.metabase.com/docs/master/developers-guide/contributing.html#yolo-just-submit-a-pr) style, but recent activity in this issue made me decide to share what I've got.

### TLDR
- I'm implementing solution from last section. Yolo PR will be send shortly.
- `approx_percentile_cont` solution is more "idiomatic" with regards to the codebase, but available only for sqlserver 2022.

### On mdx

> The SQL Server link I posted above is actually for MDX rather than normal SQL and I don't really have a good sense for whether or not that means we'd be able to leverage it.

I also have no experience working with `mdx` so while looking at this issue for the first time, I've done bit of search and found following [stackexchange answer](https://dba.stackexchange.com/a/138314), saying:

> MDX and SQL are in no way the same, and often not even comparable, as they are querying multidimensional and relational databases respectively. You cannot query your existing relational database with MDX.

This, in combination with __finding no references__ to `mdx` in your codebase made me decide that this is currently not the way to go.

### On use of [APPROX_PERCENTILE_DISC](https://learn.microsoft.com/en-us/sql/t-sql/functions/approx-percentile-disc-transact-sql?view=sql-server-ver16)

> ...and some aggregate versions like [APPROX_PERCENTILE_DISC](https://learn.microsoft.com/en-us/sql/t-sql/functions/approx-percentile-disc-transact-sql?view=sql-server-ver16)

Yes, this is aggregate function that one would use to implement this feature to fit in codebase in a way of #12096. Unfortunately `:sqlserver` has not supported this function prior to 2022 edition, which is GA for like a month now. So implementing the feature with use of this function would require at least modification of [your CI](https://github.com/metabase/metabase/blob/9eae0fc4d7c7169413972cdb6664b3788b1b0610/.github/workflows/drivers.yml#L349) to use image of `:sqlserver` 2022 for testing and flagging it to be available only to edititons 2022 and higher.

Even though it is still not widely used (imo) and with former said, I would expect this to be your desired implementation.

#### Sidenote on `:vertica`
It was some time ago when I was looking into it, but from top of my head, they also have some approx percentile function, which is supported on a newer release of the db. It think there is it just +1 major differece comparing to `:vertica` version used in [your CI](https://github.com/metabase/metabase/blob/1044a01d117303cb1a78efd628056546c87cf875/.github/workflows/drivers.yml#L372), which is still quite old.

I'will check that later today and move this info in appropriate issue.

#### Sidenote on syntax
> But the syntax looks a little wierd: `APPROX_PERCENTILE_DISC (numeric_literal)  WITHIN GROUP (ORDER BY order_by_expression [ASC|DESC]`

Very similar syntax for _aggregate percentile_ is also used by `:postgres`. `sbelak`'s implementation is taking advantage of that in [honeysql_extensions.clj](https://github.com/metabase/metabase/blob/9eae0fc4d7c7169413972cdb6664b3788b1b0610/src/metabase/util/honeysql_extensions.clj#L43) Hence implementation regarding `mbql` translation is in a way already there in the codebase. 

But there is also other way of imlpementing the feature, bit hacky, but not constrained by version. 

### Implementation with use of window function

Finally, percentile aggregation can be computed as follows.

1. Move all source impacting code from query into subquery (in sql terms `from`, `join`'s, `where`).
2. Compute percentile for every row with use of window function in subquery.
3. Use that subquery as a source for rest of aggregations, while doing `group by` also on percentile fields.

Even though, as mentioned before, your desired implementation would be imo the one from section on `approx_percentile_cont`, I've decided to try to implement this way of solving the issue. That is because I find it more interesting to solve than `approx_percentile_cont` and in case you are not interested in integrating it, you will at least have a reference PR on trying to implement aggregate via window fn.
